chain-impl-mockchain/tests/common/tx_builder.rs

```diff
@@ -25,25 +26,22 @@ impl TransactionBuilder {
         }
     }
 
-    pub fn with_input<'a>(&'a mut self, input: Input) -> &'a mut TransactionBuilder {
+    pub fn with_input<'a>(&'a mut self, input: Input) -> &'a mut Self {
         self.inputs.push(input);
         self
     }
 
-    pub fn with_output<'a>(&'a mut self, output: Output<Address>) -> &'a mut TransactionBuilder {
+    pub fn with_output<'a>(&'a mut self, output: Output<Address>) -> &'a mut Self {
         self.outputs.push(output);
         self
     }
 
-    pub fn with_outputs<'a>(
-        &'a mut self,
-        outputs: Vec<Output<Address>>,
-    ) -> &'a mut TransactionBuilder {
+    pub fn with_outputs<'a>(&'a mut self, outputs: Vec<Output<Address>>) -> &'a mut Self {
         self.outputs.extend(outputs.iter().cloned());
         self
     }
 
-    pub fn finalize<'a>(&'a mut self) -> &'a mut TransactionBuilder {
+    pub fn finalize<'a>(&'a mut self) -> &'a mut Self {
```

Comment:
This seems like it leaves the builder open for further modification, while adding the result members alongside inputs and outputs that may fall out of sync later.

I would change the method to return `Transaction` by value and remove the internal members `transaction` and `transaction_id`.

Reply:
I see now why; the final method is not actually this one, but `seal`.

To make it less confusing, it may be better to have a two-stage linear-typish builder design, where this method could be changed to

```rust
    pub fn transaction(&mut self) -> TransactionAuthenticator {
        let transaction = Transaction {
            inputs: self.inputs.clone(),
            outputs: self.outputs.clone(),
            extra: NoExtra,
        };
        TransactionAuthenticator {
            transaction,
            witnesses: Vec::new(),
            // ...
        }
    }
```

and then move the witness stuff and the `seal` method, renamed `finalize`, to `AuthenticatedTransactionBuilder`.

This may not be as nicely chainable as `&mut self` methods on the single instance, but it could be made chainable all the way through by all modifier methods changed to take `self` by value and return a modified moved value in `Self`. In test code, we should care more about convenience than performance.